include/deal.II/lac/block_linear_operator.h

```diff
@@ -162,81 +508,40 @@ block_operator(const std::array<std::array<LinearOperator<typename Range::BlockT
  *
  * @ingroup LAOperators
  */
-template <unsigned int m,
-          typename Range = BlockVector<double>,
-          typename Domain = Range>
-LinearOperator<Range, Domain>
+template <size_t m, typename Range, typename Domain>
+BlockLinearOperator<Range, Domain>
 block_diagonal_operator(const std::array<LinearOperator<typename Range::BlockType, typename Domain::BlockType>, m> &ops)
 {
   static_assert(m > 0,
                 "a blockdiagonal LinearOperator must consist of at least one block");
 
-  LinearOperator<Range, Domain> return_op;
-
-  return_op.reinit_range_vector = [ops](Range &v, bool fast)
-  {
-    // Reinitialize the block vector to m blocks:
-    v.reinit(m);
-
-    // And reinitialize every individual block with reinit_range_vectors:
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].reinit_range_vector(v.block(i), fast);
-
-    v.collect_sizes();
-  };
-
-  return_op.reinit_domain_vector = [ops](Domain &v, bool fast)
-  {
-    // Reinitialize the block vector to m blocks:
-    v.reinit(m);
-
-    // And reinitialize every individual block with reinit_domain_vectors:
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].reinit_domain_vector(v.block(i), fast);
-
-    v.collect_sizes();
-  };
-
-  return_op.vmult = [ops](Range &v, const Domain &u)
-  {
-    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
-    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
-
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].vmult(v.block(i), u.block(i));
-  };
-
-  return_op.vmult_add = [ops](Range &v, const Domain &u)
-  {
-    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
-    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
-
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].vmult_add(v.block(i), u.block(i));
-  };
-
-  return_op.Tvmult = [ops](Domain &v, const Range &u)
-  {
-    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
-    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
-
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].Tvmult(v.block(i), u.block(i));
-  };
-
-  return_op.Tvmult_add = [ops](Domain &v, const Range &u)
-  {
-    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
-    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
-
-    for (unsigned int i = 0; i < m; ++i)
-      ops[i].Tvmult_add(v.block(i), u.block(i));
-  };
-
-  return return_op;
+  typedef LinearOperator<typename Range::BlockType, typename Domain::BlockType> BlockType;
+  std::array<std::array<BlockType, m>, m> new_ops;
```

Comment:
I had to do a double take here because I thought for a moment that `BlockType` refers to either the vector's or the matrix's block types. But it's really just an operator object for individual objects. How about name it `OperatorBlock`?